File: glean-core/android/src/test/java/mozilla/telemetry/glean/private/TimingDistributionMetricTypeTest.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
* file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package mozilla.telemetry.glean.private

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import mozilla.telemetry.glean.Glean
import mozilla.telemetry.glean.GleanBuildInfo
import mozilla.telemetry.glean.GleanTimerId
import mozilla.telemetry.glean.testing.ErrorType
import mozilla.telemetry.glean.testing.GleanTestRule
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import java.lang.NullPointerException

@RunWith(AndroidJUnit4::class)
class TimingDistributionMetricTypeTest {
    val context: Context
        get() = ApplicationProvider.getApplicationContext()

    @get:Rule
    val gleanRule = GleanTestRule(ApplicationProvider.getApplicationContext())

    @Test
    fun `The API saves to its storage engine`() {
        // Define a timing distribution metric which will be stored in "store1"
        val metric = TimingDistributionMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.PING,
                name = "timing_distribution",
                sendInPings = listOf("store1"),
            ),
            timeUnit = TimeUnit.NANOSECOND,
        )

        // Accumulate a few values
        for (i in 1L..3L) {
            val id = metric.start()
            metric.stopAndAccumulate(id)
        }

        // Check that data was properly recorded.
        val snapshot = metric.testGetValue()!!
        // Check the sum
        assertTrue(snapshot.sum > 0L)
        assertEquals(snapshot.count, 3L)
        // Check that the 1L fell into the first bucket (max 1)
        // assertEquals(1L, snapshot.values[1])
        // Check that the 2L fell into the second bucket (max 2)
        // assertEquals(1L, snapshot.values[2])
        // Check that the 3L fell into the third bucket (max 3)
        // assertEquals(1L, snapshot.values[3])
    }

    @Test
    fun `disabled timing distributions must not record data`() {
        // Define a timing distribution metric which will be stored in "store1"
        // It's lifetime is set to Lifetime.PING SO IT SHOULD NOT RECORD ANYTHING.
        val metric = TimingDistributionMetricType(
            CommonMetricData(
                disabled = true,
                category = "telemetry",
                lifetime = Lifetime.PING,
                name = "timing_distribution",
                sendInPings = listOf("store1"),
            ),
            timeUnit = TimeUnit.NANOSECOND,
        )

        val id = metric.start()
        metric.stopAndAccumulate(id)

        // Check that nothing was recorded.
        assertNull(
            "Disabled TimingDistributions should not record data.",
            metric.testGetValue(),
        )
    }

    @Test
    fun `testGetValue() returns null if nothing is stored`() {
        // Define a timing distribution metric which will be stored in "store1"
        val metric = TimingDistributionMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.PING,
                name = "timing_distribution",
                sendInPings = listOf("store1"),
            ),
            timeUnit = TimeUnit.NANOSECOND,
        )
        assertNull(metric.testGetValue())
    }

    @Test
    fun `The API saves to secondary pings`() {
        // Define a timing distribution metric which will be stored in multiple stores
        val metric = TimingDistributionMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.PING,
                name = "timing_distribution",
                sendInPings = listOf("store1", "store2", "store3"),
            ),
            timeUnit = TimeUnit.NANOSECOND,
        )

        // Accumulate a few values
        for (i in 1L..3L) {
            val id = metric.start()
            metric.stopAndAccumulate(id)
        }

        // Check that data was properly recorded in the second ping.
        val snapshot = metric.testGetValue("store2")!!
        // Check the sum
        assertTrue(snapshot.sum > 0)
        assertEquals(snapshot.count, 3L)
        // Check that the 1L fell into the first bucket
        // assertEquals(1L, snapshot.values[1])
        // Check that the 2L fell into the second bucket
        // assertEquals(1L, snapshot.values[2])
        // Check that the 3L fell into the third bucket
        // assertEquals(1L, snapshot.values[3])

        // Check that data was properly recorded in the third ping.
        val snapshot2 = metric.testGetValue("store3")!!
        // Check the sum
        assertEquals(snapshot.sum, snapshot2.sum)
        assertEquals(snapshot2.count, 3L)
        // Check that the 1L fell into the first bucket
        // assertEquals(1L, snapshot2.values[1])
        // Check that the 2L fell into the second bucket
        // assertEquals(1L, snapshot2.values[2])
        // Check that the 3L fell into the third bucket
        // assertEquals(1L, snapshot2.values[3])
    }

    @Test
    fun `The accumulateSamples APIs correctly store timing values`() {
        // Define a timing distribution metric
        val metric = TimingDistributionMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.PING,
                name = "timing_distribution_samples",
                sendInPings = listOf("store1"),
            ),
            timeUnit = TimeUnit.SECOND,
        )

        // Accumulate a few values
        val testSamples = (1L..3L).toList()
        metric.accumulateSamples(testSamples)

        // Check that data was properly recorded in the second ping.
        val snapshot = metric.testGetValue("store1")!!
        val secondsToNanos = 1000L * 1000L * 1000L
        // Check the sum
        assertEquals(6L * secondsToNanos, snapshot.sum)

        // Check that we got the right number of samples.
        assertEquals(snapshot.count, 3L)

        // We should get a sample in 3 buckets.
        // These numbers are a bit magic, but they correspond to
        // `hist.sample_to_bucket_minimum(i * seconds_to_nanos)` for `i = 1..=3`,
        // which lives in the Rust code.
        assertEquals(1L, snapshot.values[984625593])
        assertEquals(1L, snapshot.values[1969251187])
        assertEquals(1L, snapshot.values[2784941737])

        // Assure the single sample API properly records.
        metric.accumulateSingleSample(4L)

        // Check that this new data was properly recorded in the second ping.
        val snapshotTwo = metric.testGetValue("store1")!!
        // Check the sum
        assertEquals(10L * secondsToNanos, snapshotTwo.sum)

        // Check that we got the right number of samples.
        assertEquals(snapshotTwo.count, 4L)
    }

    @Test
    fun `Starting a timer before initialization doesn't crash`() {
        Glean.testDestroyGleanHandle()

        val metric = TimingDistributionMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.PING,
                name = "timing_distribution_samples",
                sendInPings = listOf("store1"),
            ),
            timeUnit = TimeUnit.SECOND,
        )

        val timerId = metric.start()
        Glean.initialize(context, true, buildInfo = GleanBuildInfo.buildInfo)
        metric.stopAndAccumulate(timerId)

        assertTrue(metric.testGetValue()!!.sum >= 0)
    }

    @Test
    fun `Starting and stopping a timer before initialization doesn't crash`() {
        Glean.testDestroyGleanHandle()

        val metric = TimingDistributionMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.PING,
                name = "timing_distribution_samples",
                sendInPings = listOf("store1"),
            ),
            timeUnit = TimeUnit.SECOND,
        )

        val timerId = metric.start()
        metric.stopAndAccumulate(timerId)
        Glean.initialize(context, true, buildInfo = GleanBuildInfo.buildInfo)

        assertTrue(metric.testGetValue()!!.sum >= 0)
    }

    @Test
    fun `Stopping a non-existent timer records an error`() {
        val metric = TimingDistributionMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.PING,
                name = "timing_distribution_samples",
                sendInPings = listOf("store1"),
            ),
            timeUnit = TimeUnit.SECOND,
        )

        metric.stopAndAccumulate(GleanTimerId(1337UL))
        assertEquals(1, metric.testGetNumRecordedErrors(ErrorType.INVALID_STATE))
    }

    @Test
    fun `measure function correctly measures values`() {
        val metric = TimingDistributionMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.PING,
                name = "timing_distribution_samples",
                sendInPings = listOf("store1"),
            ),
            timeUnit = TimeUnit.NANOSECOND,
        )

        // Create a test function to "measure". This works by mocking the getElapsedNanos return
        // value setting it to return a known value to make it easier to validate.
        fun testFunc(value: Long): Long = value

        // Accumulate a few values
        for (i in 1L..3L) {
            // Measure the test function, capturing the value to verify we correctly return the
            // value of the underlying function.
            val testValue = metric.measure {
                testFunc(i)
            }

            assertEquals("Returned value must match", i, testValue)
        }

        // Check that data was properly recorded.
        val snapshot = metric.testGetValue()!!
        // Check the sum
        assertTrue(snapshot.sum > 0L)
        assertEquals(snapshot.count, 3L)
        // Check that the 1L fell into the first bucket (max 1)
        // assertEquals(1L, snapshot.values[1])
        // Check that the 2L fell into the second bucket (max 2)
        // assertEquals(1L, snapshot.values[2])
        // Check that the 3L fell into the third bucket (max 3)
        // assertEquals(1L, snapshot.values[3])
    }

    @Test
    fun `measure function does not change behavior with early return`() {
        val metric = TimingDistributionMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.PING,
                name = "inlined",
                sendInPings = listOf("store1"),
            ),
            timeUnit = TimeUnit.NANOSECOND,
        )

        // We define a function that measures the whole function call runtime
        fun testFunc(): Long =
            metric.measure {
                // We want to simulate an early return.
                if (true) {
                    // Blank 'return' is not allowed here, because `measure` is not inlined.
                    // We can return by label though.
                    return@measure 17
                }

                42
            }

        val res = testFunc()
        assertEquals("Test value must match", 17, res)

        val snapshot = metric.testGetValue()!!
        assertTrue("Should have stored some nanoseconds", snapshot.sum > 0L)

        assertEquals(snapshot.count, 1L)
    }

    @Test
    fun `measure function bubbles up exceptions and timing is canceled`() {
        val metric = TimingDistributionMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.PING,
                name = "timing_distribution_samples",
                sendInPings = listOf("store1"),
            ),
            timeUnit = TimeUnit.SECOND,
        )

        // Create a test function that throws a NPE
        fun testFunc(): Unit = throw NullPointerException()

        // Attempt to measure the function that will throw an exception.  The `measure` function
        // should allow the exception to bubble up, the timespan measurement is canceled.
        try {
            metric.measure {
                testFunc()
            }
        } catch (e: Exception) {
            // Ensure that the exception was a NPE
            assertTrue("Exception type must match", e is NullPointerException)
        } finally {
            // Check that no data was recorded.
            assertNull("Metric must not have a value", metric.testGetValue())
        }
    }
}

File: glean-core/csharp/GleanTests/Metrics/TimingDistributionMetricTypeTest.cs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

using Mozilla.Glean.Private;
using Mozilla.Glean.Testing;
using Mozilla.Glean.Utils;
using System;
using System.Collections.Generic;
using System.IO;
using Xunit;
using static Mozilla.Glean.Glean;

namespace Mozilla.Glean.Tests.Metrics
{
    public class TimingDistributionMetricTypeTest
    {
        public TimingDistributionMetricTypeTest()
        {
            // Get a random test directory just for this single test.
            string tempDataDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());

            // Make sure to clear the HighPrecisionTimestamp mocked value between tests.
            HighPrecisionTimestamp.MockedValue = null;

            // In xUnit, the constructor will be called before each test. This
            // feels like a natural place to initialize / reset Glean.
            GleanInstance.Reset(
	            applicationId: "org.mozilla.csharp.tests",
                applicationVersion: "1.0-test",
                uploadEnabled: true,
                configuration: new Configuration(),
                dataDir: tempDataDir
                );
        }

        [Fact]
        public void APISavesToStorage()
        {
            TimingDistributionMetricType metric = new TimingDistributionMetricType(
                disabled: false,
                category: "telemetry",
                lifetime: Lifetime.Ping,
                name: "timing_distribution",
                sendInPings: new string[] { "store1" },
                timeUnit: TimeUnit.Nanosecond
            );

            // Accumulate a few values
            for (ulong i = 1; i <= 3; i++)
            {
                HighPrecisionTimestamp.MockedValue = 0;
                GleanTimerId id = metric.Start();
                HighPrecisionTimestamp.MockedValue = i;
                metric.StopAndAccumulate(id);
            }

            // Check that data was properly recorded.
            Assert.True(metric.TestHasValue());
            var snapshot = metric.TestGetValue();
            // Check the sum
            Assert.Equal(6L, snapshot.Sum);
            // Check that the 1L fell into the first bucket (max 1)
            Assert.Equal(1L, snapshot.Values[1]);
            // Check that the 2L fell into the second bucket (max 2)
            Assert.Equal(1L, snapshot.Values[2]);
            // Check that the 3L fell into the third bucket (max 3)
            Assert.Equal(1L, snapshot.Values[3]);
        }

        [Fact]
        public void DisableTimingDistributionsMustNotRecordData()
        {
            TimingDistributionMetricType metric = new TimingDistributionMetricType(
                disabled: true,
                category: "telemetry",
                lifetime: Lifetime.Ping,
                name: "timing_distribution",
                sendInPings: new string[] { "store1" },
                timeUnit: TimeUnit.Nanosecond
            );

            // Attempt to store the timing distribution using set
            GleanTimerId id = metric.Start();
            metric.StopAndAccumulate(id);

            // Check that nothing was recorded.
            Assert.False(metric.TestHasValue(), "Disabled TimingDistributions should not record data.");
        }

        [Fact]
        public void TestGetValueThrows()
        {
            TimingDistributionMetricType metric = new TimingDistributionMetricType(
                disabled: false,
                category: "telemetry",
                lifetime: Lifetime.Ping,
                name: "timing_distribution",
                sendInPings: new string[] { "store1" },
                timeUnit: TimeUnit.Nanosecond
            );
            Assert.Throws<NullReferenceException>(() => metric.TestGetValue());
        }

        [Fact]
        public void APISavesToSecondaryPings()
        {
            TimingDistributionMetricType metric = new TimingDistributionMetricType(
                disabled: false,
                category: "telemetry",
                lifetime: Lifetime.Ping,
                name: "timing_distribution",
                sendInPings: new string[] { "store1", "store2", "store3" },
                timeUnit: TimeUnit.Nanosecond
            );

            // Accumulate a few values
            for (ulong i = 1; i <= 3; i++)
            {
                HighPrecisionTimestamp.MockedValue = 0;
                GleanTimerId id = metric.Start();
                HighPrecisionTimestamp.MockedValue = i;
                metric.StopAndAccumulate(id);
            }

            // Check that data was properly recorded in the second ping.
            Assert.True(metric.TestHasValue("store2"));
            var snapshot = metric.TestGetValue("store2");
            // Check the sum
            Assert.Equal(6L, snapshot.Sum);
            // Check that the 1L fell into the first bucket
            Assert.Equal(1L, snapshot.Values[1]);
            // Check that the 2L fell into the second bucket
            Assert.Equal(1L, snapshot.Values[2]);
            // Check that the 3L fell into the third bucket
            Assert.Equal(1L, snapshot.Values[3]);

            // Check that data was properly recorded in the third ping.
            Assert.True(metric.TestHasValue("store3"));
            var snapshot2 = metric.TestGetValue("store3");
            // Check the sum
            Assert.Equal(6L, snapshot2.Sum);
            // Check that the 1L fell into the first bucket
            Assert.Equal(1L, snapshot2.Values[1]);
            // Check that the 2L fell into the second bucket
            Assert.Equal(1L, snapshot2.Values[2]);
            // Check that the 3L fell into the third bucket
            Assert.Equal(1L, snapshot2.Values[3]);
        }

        [Fact]
        public void StartingATimerBeforeInitializationDoesNotCrash()
        {
            GleanInstance.TestDestroyGleanHandle();
            Dispatchers.QueueInitialTasks = true;

            TimingDistributionMetricType metric = new TimingDistributionMetricType(
                disabled: false,
                category: "telemetry",
                lifetime: Lifetime.Ping,
                name: "timing_distribution",
                sendInPings: new string[] { "store1" },
                timeUnit: TimeUnit.Second
            );

            GleanTimerId timerId = metric.Start();

            // Start Glean again.
            string tempDataDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
            GleanInstance.Initialize(
                applicationId: "org.mozilla.csharp.tests",
                applicationVersion: "1.0-test",
                uploadEnabled: true,
                configuration: new Configuration(),
                dataDir: tempDataDir
            );

            metric.StopAndAccumulate(timerId);
            Assert.True(metric.TestGetValue().Sum >= 0);
        }

        [Fact]
        public void StartingAndStoppingATimerBeforeInitializationDoesNotCrash()
        {
            GleanInstance.TestDestroyGleanHandle();
            Dispatchers.QueueInitialTasks = true;

            TimingDistributionMetricType metric = new TimingDistributionMetricType(
                disabled: false,
                category: "telemetry",
                lifetime: Lifetime.Ping,
                name: "timing_distribution",
                sendInPings: new string[] { "store1" },
                timeUnit: TimeUnit.Second
            );

            GleanTimerId timerId = metric.Start();
            metric.StopAndAccumulate(timerId);

            // Start Glean again.
            string tempDataDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
            GleanInstance.Initialize(
                applicationId: "org.mozilla.csharp.tests",
                applicationVersion: "1.0-test",
                uploadEnabled: true,
                configuration: new Configuration(),
                dataDir: tempDataDir
            );
            Assert.True(metric.TestGetValue().Sum >= 0);
        }

        [Fact]
        public void StoppingANonExistentTimerRecordsAnError()
        {
            TimingDistributionMetricType metric = new TimingDistributionMetricType(
                disabled: false,
                category: "telemetry",
                lifetime: Lifetime.Ping,
                name: "timing_distribution",
                sendInPings: new string[] { "store1" },
                timeUnit: TimeUnit.Second
            );

            // Hopefully ulong.MaxValue wasn't used already :)
            metric.StopAndAccumulate((GleanTimerId)(ulong.MaxValue));
            Assert.Equal(1, metric.TestGetNumRecordedErrors(ErrorType.InvalidState));
        }

        [Fact]
        public void MeasureFunctionCorrectlyMeasuresValues()
        {
            TimingDistributionMetricType metric = new TimingDistributionMetricType(
                disabled: false,
                category: "telemetry",
                lifetime: Lifetime.Ping,
                name: "timing_distribution",
                sendInPings: new string[] { "store1" },
                timeUnit: TimeUnit.Nanosecond
            );

            // Create a test function to "measure". This works by mocking the timer return
            // value setting it to return a known value to make it easier to validate.
            static ulong TestFunc(ulong value) {
                HighPrecisionTimestamp.MockedValue = value;
                return value;
            }

            // Accumulate a few values
            for (ulong i = 1; i <= 3; i++)
            {
                // Measure the test function, capturing the value to verify we correctly return the
                // value of the underlying function.
                HighPrecisionTimestamp.MockedValue = 0;
                var testValue = metric.Measure(() => TestFunc(i));

                // Returned value must match
                Assert.Equal(i, testValue);
            }

            // Check that data was properly recorded.
            Assert.True(metric.TestHasValue());
            var snapshot = metric.TestGetValue();
            // Check the sum
            Assert.Equal(6L, snapshot.Sum);
            // Check that the 1L fell into the first bucket (max 1)
            Assert.Equal(1L, snapshot.Values[1]);
            // Check that the 2L fell into the second bucket (max 2)
            Assert.Equal(1L, snapshot.Values[2]);
            // Check that the 3L fell into the third bucket (max 3)
            Assert.Equal(1L, snapshot.Values[3]);
        }

        [Fact]
        public void MeasureFunctionBubblesUpExceptionsAndTimingIsCanceled()
        {
            TimingDistributionMetricType metric = new TimingDistributionMetricType(
                disabled: false,
                category: "telemetry",
                lifetime: Lifetime.Ping,
                name: "timing_distribution",
                sendInPings: new string[] { "store1" },
                timeUnit: TimeUnit.Second
            );

            // Create a test function that throws a NPE
            static bool TestFunc()
            {
                throw new NullReferenceException();
            }

            // Attempt to measure the function that will throw an exception.  The `measure` function
            // should allow the exception to bubble up, the timing distribution measurement is canceled.
            Assert.Throws<NullReferenceException>(() => metric.Measure(() => TestFunc()));
        }

        [Fact]
        public void EnsureThatTimeUnitControlsTruncation()
        {
            ulong maxSampleTime = 1000L * 1000 * 1000 * 60 * 10;

            foreach (TimeUnit unit in new List<TimeUnit>() {
                    TimeUnit.Nanosecond,
                    TimeUnit.Microsecond,
                    TimeUnit.Millisecond
            })
            {
                TimingDistributionMetricType metric = new TimingDistributionMetricType(
                    disabled: false,
                    category: "telemetry",
                    lifetime: Lifetime.Ping,
                    name: $"test_{unit}",
                    sendInPings: new string[] { "store1" },
                    timeUnit: unit
                );

                foreach (ulong value in new List<ulong>() {
                    1L,
                    1000L,
                    100000L,
                    maxSampleTime,
                    maxSampleTime* 1000L,
                    maxSampleTime* 1000000L
                })
                {
                    HighPrecisionTimestamp.MockedValue = 0;
                    var timerId = metric.Start();
                    HighPrecisionTimestamp.MockedValue = value;
                    metric.StopAndAccumulate(timerId);
                }

                var snapshot = metric.TestGetValue();
                Assert.True(snapshot.Values.Count < 318);
            }
        }
    }
}
